code-formatting/pre-commit.sh

```
@@ -21,8 +83,22 @@ STAGED_JAVA_FILES_AS_REGEX=$(git diff --staged --name-only --diff-filter=ACMR |
+   # Spotless must run in the Maven modules that own the staged files. Running `spotless:apply` at
+   # the aggregator root fails ("No plugin found for prefix 'spotless'" — the root only manages the
+   # plugin), and the fully-qualified goal errors on modules that don't configure Spotless (e.g. the
+   # BOM). So derive each staged file's owning module (nearest ancestor with a pom.xml) and pass them
+   # via -pl. -DspotlessFiles still restricts formatting to exactly the staged files within them.
+   SPOTLESS_MODULES=$(git diff --staged --name-only --diff-filter=ACMR | grep '.java$' | while read -r f; do
+       d=$(dirname "$f")
+       while [ "$d" != "." ] && [ ! -f "$d/pom.xml" ]; do d=$(dirname "$d"); done
+       [ -f "$d/pom.xml" ] && printf '%s\n' "$d"
+   done | sort -u | paste -sd',' -)
+   if [ -n "$SPOTLESS_MODULES" ]; then
+       mvn -pl "$SPOTLESS_MODULES" spotless:apply -DspotlessFiles=^.*$STAGED_JAVA_FILES_AS_REGEX
```

Shellcheck (0.11.0)

[info] Double quote to prevent globbing and word splitting.

(SC2086)